From: Vorapol Rinsatitnon <vorapol.r@pm.me>
Date: Fri, 21 Aug 2026 17:04:08 +1000
Subject: [PATCH] Implement atomic And/Or via CAS under race

Legacy race_windows.syso lacks __tsan_go_atomic*_fetch_and/or.
Provide race-tagged Go implementations using Load+CompareAndSwap,
and keep asm-backed declarations under !race only.

diff --git a/src/sync/atomic/andor_decl.go b/src/sync/atomic/andor_decl.go
new file mode 100644
index 00000000..c6cd2203
--- /dev/null
+++ b/src/sync/atomic/andor_decl.go
@@ -0,0 +1,49 @@
+// Copyright 2026 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build !race
+
+package atomic
+
+// AndInt32 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Int32.And] instead.
+//
+//go:noescape
+func AndInt32(addr *int32, mask int32) (old int32)
+
+// AndUint32 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Uint32.And] instead.
+//
+//go:noescape
+func AndUint32(addr *uint32, mask uint32) (old uint32)
+
+// AndUintptr atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Uintptr.And] instead.
+//
+//go:noescape
+func AndUintptr(addr *uintptr, mask uintptr) (old uintptr)
+
+// OrInt32 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Int32.Or] instead.
+//
+//go:noescape
+func OrInt32(addr *int32, mask int32) (old int32)
+
+// OrUint32 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Uint32.Or] instead.
+//
+//go:noescape
+func OrUint32(addr *uint32, mask uint32) (old uint32)
+
+// OrUintptr atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Uintptr.Or] instead.
+//
+//go:noescape
+func OrUintptr(addr *uintptr, mask uintptr) (old uintptr)
diff --git a/src/sync/atomic/andor_decl_32.go b/src/sync/atomic/andor_decl_32.go
new file mode 100644
index 00000000..b76d221b
--- /dev/null
+++ b/src/sync/atomic/andor_decl_32.go
@@ -0,0 +1,27 @@
+// Copyright 2026 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build !race && (386 || arm || mips || mipsle)
+
+package atomic
+
+// AndInt64 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Int64.And] instead.
+func AndInt64(addr *int64, mask int64) (old int64)
+
+// AndUint64 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
+// and returns the old.
+// Consider using the more ergonomic and less error-prone [Uint64.And] instead.
+func AndUint64(addr *uint64, mask uint64) (old uint64)
+
+// OrInt64 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Int64.Or] instead.
+func OrInt64(addr *int64, mask int64) (old int64)
+
+// OrUint64 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Uint64.Or] instead.
+func OrUint64(addr *uint64, mask uint64) (old uint64)
diff --git a/src/sync/atomic/andor_decl_64.go b/src/sync/atomic/andor_decl_64.go
new file mode 100644
index 00000000..0e81dc42
--- /dev/null
+++ b/src/sync/atomic/andor_decl_64.go
@@ -0,0 +1,35 @@
+// Copyright 2026 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build !race && !(386 || arm || mips || mipsle)
+
+package atomic
+
+// AndInt64 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Int64.And] instead.
+//
+//go:noescape
+func AndInt64(addr *int64, mask int64) (old int64)
+
+// AndUint64 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
+// and returns the old.
+// Consider using the more ergonomic and less error-prone [Uint64.And] instead.
+//
+//go:noescape
+func AndUint64(addr *uint64, mask uint64) (old uint64)
+
+// OrInt64 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Int64.Or] instead.
+//
+//go:noescape
+func OrInt64(addr *int64, mask int64) (old int64)
+
+// OrUint64 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
+// and returns the old value.
+// Consider using the more ergonomic and less error-prone [Uint64.Or] instead.
+//
+//go:noescape
+func OrUint64(addr *uint64, mask uint64) (old uint64)
diff --git a/src/sync/atomic/andor_race.go b/src/sync/atomic/andor_race.go
new file mode 100644
index 00000000..65e25293
--- /dev/null
+++ b/src/sync/atomic/andor_race.go
@@ -0,0 +1,102 @@
+// Copyright 2026 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build race
+
+// And/Or under the race detector are implemented with Load+CAS loops so that
+// the legacy Windows race_windows.syso (which lacks
+// __tsan_go_atomic*_fetch_and/or) can still link. Race instrumentation comes
+// from the existing Load/CompareAndSwap TSAN hooks.
+
+package atomic
+
+func AndInt32(addr *int32, mask int32) (old int32) {
+	for {
+		old = LoadInt32(addr)
+		if CompareAndSwapInt32(addr, old, old&mask) {
+			return old
+		}
+	}
+}
+
+func AndUint32(addr *uint32, mask uint32) (old uint32) {
+	for {
+		old = LoadUint32(addr)
+		if CompareAndSwapUint32(addr, old, old&mask) {
+			return old
+		}
+	}
+}
+
+func AndInt64(addr *int64, mask int64) (old int64) {
+	for {
+		old = LoadInt64(addr)
+		if CompareAndSwapInt64(addr, old, old&mask) {
+			return old
+		}
+	}
+}
+
+func AndUint64(addr *uint64, mask uint64) (old uint64) {
+	for {
+		old = LoadUint64(addr)
+		if CompareAndSwapUint64(addr, old, old&mask) {
+			return old
+		}
+	}
+}
+
+func AndUintptr(addr *uintptr, mask uintptr) (old uintptr) {
+	for {
+		old = LoadUintptr(addr)
+		if CompareAndSwapUintptr(addr, old, old&mask) {
+			return old
+		}
+	}
+}
+
+func OrInt32(addr *int32, mask int32) (old int32) {
+	for {
+		old = LoadInt32(addr)
+		if CompareAndSwapInt32(addr, old, old|mask) {
+			return old
+		}
+	}
+}
+
+func OrUint32(addr *uint32, mask uint32) (old uint32) {
+	for {
+		old = LoadUint32(addr)
+		if CompareAndSwapUint32(addr, old, old|mask) {
+			return old
+		}
+	}
+}
+
+func OrInt64(addr *int64, mask int64) (old int64) {
+	for {
+		old = LoadInt64(addr)
+		if CompareAndSwapInt64(addr, old, old|mask) {
+			return old
+		}
+	}
+}
+
+func OrUint64(addr *uint64, mask uint64) (old uint64) {
+	for {
+		old = LoadUint64(addr)
+		if CompareAndSwapUint64(addr, old, old|mask) {
+			return old
+		}
+	}
+}
+
+func OrUintptr(addr *uintptr, mask uintptr) (old uintptr) {
+	for {
+		old = LoadUintptr(addr)
+		if CompareAndSwapUintptr(addr, old, old|mask) {
+			return old
+		}
+	}
+}
diff --git a/src/sync/atomic/doc.go b/src/sync/atomic/doc.go
index 4e934047..8e3a4f4b 100644
--- a/src/sync/atomic/doc.go
+++ b/src/sync/atomic/doc.go
@@ -128,48 +128,6 @@ func AddUint32(addr *uint32, delta uint32) (new uint32)
 //go:noescape
 func AddUintptr(addr *uintptr, delta uintptr) (new uintptr)
 
-// AndInt32 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Int32.And] instead.
-//
-//go:noescape
-func AndInt32(addr *int32, mask int32) (old int32)
-
-// AndUint32 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Uint32.And] instead.
-//
-//go:noescape
-func AndUint32(addr *uint32, mask uint32) (old uint32)
-
-// AndUintptr atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Uintptr.And] instead.
-//
-//go:noescape
-func AndUintptr(addr *uintptr, mask uintptr) (old uintptr)
-
-// OrInt32 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Int32.Or] instead.
-//
-//go:noescape
-func OrInt32(addr *int32, mask int32) (old int32)
-
-// OrUint32 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Uint32.Or] instead.
-//
-//go:noescape
-func OrUint32(addr *uint32, mask uint32) (old uint32)
-
-// OrUintptr atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Uintptr.Or] instead.
-//
-//go:noescape
-func OrUintptr(addr *uintptr, mask uintptr) (old uintptr)
-
 // LoadInt32 atomically loads *addr.
 // Consider using the more ergonomic and less error-prone [Int32.Load] instead.
 //
diff --git a/src/sync/atomic/doc_32.go b/src/sync/atomic/doc_32.go
index 9d644f25..6d7f3857 100644
--- a/src/sync/atomic/doc_32.go
+++ b/src/sync/atomic/doc_32.go
@@ -38,26 +38,6 @@ func AddInt64(addr *int64, delta int64) (new int64)
 // (particularly if you target 32-bit platforms; see the bugs section).
 func AddUint64(addr *uint64, delta uint64) (new uint64)
 
-// AndInt64 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Int64.And] instead.
-func AndInt64(addr *int64, mask int64) (old int64)
-
-// AndUint64 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
-// and returns the old.
-// Consider using the more ergonomic and less error-prone [Uint64.And] instead.
-func AndUint64(addr *uint64, mask uint64) (old uint64)
-
-// OrInt64 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Int64.Or] instead.
-func OrInt64(addr *int64, mask int64) (old int64)
-
-// OrUint64 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Uint64.Or] instead.
-func OrUint64(addr *uint64, mask uint64) (old uint64)
-
 // LoadInt64 atomically loads *addr.
 // Consider using the more ergonomic and less error-prone [Int64.Load] instead
 // (particularly if you target 32-bit platforms; see the bugs section).
diff --git a/src/sync/atomic/doc_64.go b/src/sync/atomic/doc_64.go
index 5fec3f4e..c819ba9c 100644
--- a/src/sync/atomic/doc_64.go
+++ b/src/sync/atomic/doc_64.go
@@ -50,34 +50,6 @@ func AddInt64(addr *int64, delta int64) (new int64)
 //go:noescape
 func AddUint64(addr *uint64, delta uint64) (new uint64)
 
-// AndInt64 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Int64.And] instead.
-//
-//go:noescape
-func AndInt64(addr *int64, mask int64) (old int64)
-
-// AndUint64 atomically performs a bitwise AND operation on *addr using the bitmask provided as mask
-// and returns the old.
-// Consider using the more ergonomic and less error-prone [Uint64.And] instead.
-//
-//go:noescape
-func AndUint64(addr *uint64, mask uint64) (old uint64)
-
-// OrInt64 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Int64.Or] instead.
-//
-//go:noescape
-func OrInt64(addr *int64, mask int64) (old int64)
-
-// OrUint64 atomically performs a bitwise OR operation on *addr using the bitmask provided as mask
-// and returns the old value.
-// Consider using the more ergonomic and less error-prone [Uint64.Or] instead.
-//
-//go:noescape
-func OrUint64(addr *uint64, mask uint64) (old uint64)
-
 // LoadInt64 atomically loads *addr.
 // Consider using the more ergonomic and less error-prone [Int64.Load] instead
 // (particularly if you target 32-bit platforms; see the bugs section).
